sycl/doc/CompilerAndRuntimeDesign.md

```diff
@@ -787,18 +787,61 @@ The SPIR-V specific functions are implemented in for the SYCL host device here:
 
 ### Address spaces handling
 
-SYCL 1.2.1 language defines several address spaces where data can reside, the
-same as OpenCL - global, local, private and constant. From the spec: "In OpenCL
-C, these address spaces are manually specified using OpenCL-specific keywords.
-In SYCL, the device compiler is expected to auto-deduce the address space for
-pointers in common situations of pointer usage. However, there are situations
-where auto-deduction is not possible".
+SYCL-1.2.1 specification uses C++ templates to represent pointers to disjoint
+memory regions on an accelerator to enable compilation with standard C++
+toolchain and SYCL compiler toolchain.
+
+For instance:
+
+``` C++
+// check that SYCL mode is ON and we can use non-standard annotations
+#if defined(__SYCL_DEVICE_ONLY__)
+// GPU/accelerator implementation
+template <typename T, address_space AS> class multi_ptr {
+  // GetAnnotatedPointer<T, global>::type == "__attribute__((opencl_global)) T"
+  using pointer_t = typename GetAnnotatedPointer<T, AS>::type *;
+ 
+  pointer_t data;
+  public:
+  pointer_t get_pointer() { return data; }
+}
+#else
+// CPU/host implementation
+template <typename T, address_space AS> class multi_ptr {
+  T *data; // regular unannotated pointer
+  public:
+  T *get_pointer() { return data; }
+}
+#endif
+```
 
-We believe that requirement for the compiler to automatically deduce address
-spaces is too strong. Instead the following approach will be implemented in the
-compiler:
+Depending on the compiler mode `multi_ptr` will either annotate internal data
+with address space attribute or not.
 
-*TBD*
+To utilize existing clang's functionality, we re-use following OpenCL address
+space attributes in SYCL mode with modified semantic rules and unmodified parser
+and IR generation components.
+
+``` C++
+  __attribute__((opencl_global))
+  __attribute__((opencl_local))
+  __attribute__((opencl_private))
+```
+
+The main difference in Sema component with OpenCL mode is that SYCL mode
+(similar to other single-source GPU programming modes like OpenMP/CUDA/HIP)
+keeps "default" address space for the declaration without explicit address space
+attribute annotations. This keeps the type system for the code shared between
+the host and device semantically-equivalent for both compilers: regular C++ host
+compiler and SYCL compiler.
+
+SYCL mode allows conversion from annotated pointer to unannotated pointer to
+enable integration of accelerated code with standard C++ code.
+
+> **NOTE:** casting unannotated pointer to address-space annotated pointer must
```

Comment:
Why is this casting allowed at all at source level?

Reply:
To enable efficient memory access. Access via unannotated pointer implies additional jit-time or run-time overhead.

Reply:
I mean if we state that casting (with non-matching AS is UB), then why can't we instead disallow that at all?

Reply:
We follow OpenCL semantic here (see section 6.5.5 "The generic address space" of OpenCL C v2.0 specificaton) and according to my understanding it's useful for cases when address space is known, although pointers are not annotated.

```c++

void some_user_func(int *p) {
  // p is not explicitly annotated, but user knows function is always called with a private pointer
  // cast to private
  multi_ptr<int, sycl::access::private_space> x(p);
  ...
}
```

Reply:
That's fine for "generic". I was suggesting to explicitly forbid casting from a known AS to another known AS, which is different from the first one, e.g. "private" -> "global". But I understand now that this is not the case you were discussing in your note. BTW, why you removed it in your later revision?

Reply:
> I was suggesting to explicitly forbid casting from a known AS to another known AS, which is different from the first one, e.g. "private" -> "global".

This cast is explicitly prohibited by OpenCL (section 6.5 "Address Space Qualifiers"):

> A pointer to address space A can only be assigned to a pointer to the same address space A or a
> pointer to the generic address space. Casting a pointer to address space A to a pointer to address
> space B is illegal if A and B are named address spaces and A is not the same as B.

BTW, I noticed that DPC++ doesn't detect this illegal cast in the FE, but fails to convert it to SPIR-V file:
``` C++
#include <CL/sycl.hpp>
using namespace sycl;
SYCL_EXTERNAL private_ptr<float> bar(global_ptr<float> f) {
  return private_ptr<float>{f.get()};
}
```

``` bash
InvalidModule: Invalid SPIR-V module: Casts from private/local/global address space are allowed only to generic
   %1 = addrspacecast float addrspace(1)* %0 to float*
```
I'll file an issue for that.

> BTW, why you removed it in your later revision?

I decided refer to OpenCL semantic instead of trying to describe it in my own words - this seems to be less error-prone approach. :)

Reply:
Sounds good, thanks.